### PR TITLE
Upstream SPV_INTEL_masked_gather_scatter

### PR DESCRIPTION
**Pull request by MrSidims**
This extension allows TypeVector to have a physical pointer type Component Type and introduces gather/scatter instructions. It will be useful for explicitly vectorized kernels.

SPIR-V validator adjustments will be done later.

SPIR-V Headers: https://github.com/KhronosGroup/SPIRV-Headers/pull/391

### CODE REVIEW

File: extensions/INTEL/SPV_INTEL_masked_gather_scatter.asciidoc [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+Revision History
+----------------
+[cols="5,15,15,70"]
+[grid="rows"]
+[options="header"]
+|========================================
+|Rev|Date|Author|Changes
+|1|2022-08-19|Dmitry Sidorov|Initial revision
+|2|2023-09-05|Dmitry Sidorov|Prepare to ship
```
Comment (bashbaug) [inline extensions/INTEL/SPV_INTEL_masked_gather_scatter.asciidoc:290]: The first published version is usually version 1, but I'm fine publishing version 2 if you are.

**Review by bashbaug — APPROVED** [review]
LGTM, just one minor comment regarding the extension version.

File: extensions/INTEL/SPV_INTEL_masked_gather_scatter.asciidoc [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+|=====
+
+[cols="1,1,4*3",width="100%"]
+|=====
+5+|[[OpMaskedScatterINTEL]]*OpMaskedScatterINTEL* +
+ +
+Writes values from a vector to the corresponding memory address of the given vector of pointers. Memory access
+is specified by a mask instruction parameter. +
+ +
+'InputVector' is a _vector_ of values to scatter. +
```
Comment (alan-baker) [inline extensions/INTEL/SPV_INTEL_masked_gather_scatter.asciidoc:216]: Not a blocker, but OpStore puts the pointer operand first. Not sure if you want to follow that precedent.

Comment (MrSidims) [reply extensions/INTEL/SPV_INTEL_masked_gather_scatter.asciidoc:216]: Good comment. I agree, that the order should follow OpStore. The problem though is that we have implemented the instruction across the software stack and changing order now would be backward incompatible. I see two possible ways to deal with it:
1. Change Op code and the order now, leaving for us to support both old and new Op codes and orders;
2. If somebody express an interest to make this extension EXT/KHR or add the instructions to the core spec - we can change the Op code and the order then.

I'd slightly prefer to stick with the option 2, wdyt?

Comment (alan-baker) [reply extensions/INTEL/SPV_INTEL_masked_gather_scatter.asciidoc:216]: If you've implemented it already I agree it is not worth changing now. We can think about option 2 if there is ever wider adoption.

**Pull request merged**